homeassistant/components/rainbird/__init__.py

```diff
@@ -61,47 +70,101 @@
-                _setup_controller(hass, controller_config, config)
-                for controller_config in config[DOMAIN]
-            ]
+    if DOMAIN not in config:
+        return True
+
+    for controller_config in config[DOMAIN]:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=controller_config,
+            )
         )
+
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2023.3.0",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
     )
 
+    return True
 
-async def _setup_controller(hass, controller_config, config):
-    """Set up a controller."""
-    server = controller_config[CONF_HOST]
-    password = controller_config[CONF_PASSWORD]
-    client = AsyncRainbirdClient(async_get_clientsession(hass), server, password)
-    controller = AsyncRainbirdController(client)
-    try:
-        await controller.get_serial_number()
-    except RainbirdApiException as exc:
-        _LOGGER.error("Unable to setup controller: %s", exc)
-        return False
 
-    rain_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_sensor_state)
-    delay_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_delay)
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the config entry for Rain Bird."""
+    hass.data.setdefault(DOMAIN, {})
 
-    for platform in PLATFORMS:
-        hass.async_create_task(
-            discovery.async_load_platform(
-                hass,
-                platform,
-                DOMAIN,
-                {
-                    RAINBIRD_CONTROLLER: controller,
-                    SENSOR_TYPE_RAINSENSOR: rain_coordinator,
-                    SENSOR_TYPE_RAINDELAY: delay_coordinator,
-                    **controller_config,
-                },
-                config,
-            )
+    controller = AsyncRainbirdController(
+        AsyncRainbirdClient(
+            async_get_clientsession(hass),
+            entry.data[CONF_HOST],
+            entry.data[CONF_PASSWORD],
         )
+    )
+
+    try:
+        async with async_timeout.timeout(TIMEOUT_SECONDS):
+            serial_number = await controller.get_serial_number()
+    except (RainbirdApiException, asyncio.TimeoutError) as err:
+        raise ConfigEntryNotReady(f"Error talking to controller: {str(err)}") from err
+
+    device_info = DeviceInfo(
+        default_name=MANUFACTURER,
+        identifiers={(DOMAIN, serial_number)},
+        manufacturer=MANUFACTURER,
+    )
+    rain_coordinator = RainbirdUpdateCoordinator(
+        hass, "Rain", controller.get_rain_sensor_state
+    )
+    delay_coordinator = RainbirdUpdateCoordinator(
+        hass, "Rain delay", controller.get_rain_delay
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = {
```

Comment:
Based on this comment and the comment about initializing the update coordinators in the platforms, I decided to just simplify the overall interactions between the device and the integration, moving all the rpcs into a single data update coordinator. This required changing the tests, so i added additional test fixtures to help manage the request setup and ordering expected by the integration.

The consequence of this is now a single update coordinator handles rpcs for all platforms so there isn't independence for each sensor, but in hindsight that is probably simpler overall (e.g. if the device is offline, there is just one coordinator trying to hit it rather than 3). 

While I was here, I also removed some unnecessary duplicate sensors. Today there is a rain sensor and a rain delay sensor both exported as a sensor and binary sensor and it doesn't really make a lot of sense to me. I've just left the sensors that provide unique value and have updated the breaking changes in the description.

Reply:
If this is too much for a single PR, I can break into smaller chunks, or move into another branch to not have to worry about leaving dev in a bad state.

Reply:
It's not too much.

What did the four binary sensors and sensors represent? How would a user replace the removed sensors?

Reply:
There were four total sensors representing two values that were redundant. I updated the PR to describe but can reword if it is not clear.

The "rain" sensor was a binary sensor and a sensor with a string boolean value. This represents whether the device knows it is raining (I'm not sure if it's from a separate sensors only or also from a cloud weather report)

The "rain delay" represents the number of days that the sprinkler is paused due to rain.  This is a sensor and also had a binary sensor when it was non-zero.  The rain delay service can update this value so maybe this should just be a Number?

Reply:
Ok.

Yeah, if the rain delay service and sensor are in sync it should be a number entity.

Reply:
We can deprecate the service and raise a repair issue telling users to move to the number entity and service when that's in place.